contracts/CliptoExchange.sol

```diff
@@ -0,0 +1,242 @@
+    event CreatorRegistered(address indexed creator, address indexed nft, string jsondata);
+    event CreatorUpdated(address indexed creator, string jsondata);
+    event NewRequest(address indexed creator, uint256 requestId, string jsondata);
+    event DeliveredRequest(address indexed creator, uint256 requestId, uint256 nftTokenId);
+    event RefundedRequest(address indexed creator, uint256 requestId);
+    event RejectRequest(address indexed creator, uint256 requestId);
+    event MigrationCreator(address[] creators);
+
+    function initialize(address _feeRecipient, address _beacon) public initializer {
+        __ReentrancyGuard_init();
+        __Pausable_init();
+
+        beacon = _beacon;
+        owner = msg.sender;
+        feeRecipient = _feeRecipient;
+        _feeDenom = 1;
+    }
+
+    function getRequest(address _creator, uint256 _requestId) external view returns (Request memory) {
+        return requests[_creator][_requestId];
+    }
+
+    function getCreator(address _creator) external view returns (address) {
+        return creators[_creator];
+    }
+
+    function getFeeRate() external view returns (uint256, uint256) {
+        return (_feeNumer, _feeDenom);
+    }
+
+    function setFeeRate(uint256 feeNumer_, uint256 feeDenom_) external onlyOwner {
+        require(feeDenom_ != 0, "error: denom should be non zero");
+        require(feeNumer_ <= feeDenom_, "error: denom should be smaller than numer");
+
+        _feeNumer = feeNumer_;
+        _feeDenom = feeDenom_;
+    }
+
+    function setFeeRecipient(address _feeRecipient) external onlyOwner {
+        feeRecipient = _feeRecipient;
+    }
+
+    function transferOwnership(address newOwner) external onlyOwner {
+        address oldOwner = owner;
+        owner = newOwner;
+        emit OwnershipTransferred(oldOwner, newOwner);
+    }
+
+    function setCliptoMinter(address _cliptoToken, address _minter) external onlyOwner {
+        ICliptoToken(_cliptoToken).setMinter(_minter);
+    }
+
+    function registerCreator(string calldata _creatorName, string calldata _jsondata) external whenNotPaused {
+        require(!_existsCreator(msg.sender), "error: creator already registered");
+        _registerCreator(msg.sender, _creatorName, _jsondata);
+    }
+
+    function updateCreator(string calldata _jsondata) external whenNotPaused {
+        require(_existsCreator(msg.sender), "error: creator is not yet registered");
+        emit CreatorUpdated(msg.sender, _jsondata);
+    }
+
+    function newRequest(
+        address _creator,
+        address _nftReceiver,
+        address _erc20,
+        uint256 _amount,
+        string calldata _jsonData
+    ) external whenNotPaused {
+        _validateRequest(_creator, _amount);
+        _checkAllowance(_erc20, msg.sender, _amount);
+        _pay(msg.sender, address(this), _erc20, _amount);
+        _newRequest(_creator, _nftReceiver, _erc20, _amount, _jsonData);
```

Comment:
If we're going to accept arbitrary ERC20s as specified by users as payment, there should be big red warnings for creators if the token isn't within a token list (a-la Uniswap token list implementation/etc)

Reply:
We'll also need to manually blacklist things like rebasing tokens, tokens which have taxes on transfers, reflections, etc, as these will be incompatible and cause really bad UX. Maybe even a whitelist tbh.

Reply:
we would be filtering out the requests in the FE if the token used for the request is not in our list of acceptable tokens.
Creators will not see these request but requester can claim refund once the deadline is passed